### PR TITLE
Small improvements to make genotyping code more readable.

### LINKED ISSUE #1684: Code duplication in InfiniteRandomMatingPopulationModel
In `InfiniteRandomMatingPopulationModel` the methods `singleSampleLikelihoods`, `multiSampleHeterogeneousPloidyModelLikelihoods`, and `multiSampleHomogeneousPloidyModelLikelihoods` do almost exactly the same thing with almost identical code.  It does not appear that the assumptions of a single sample or homogeneous ploidy are used for any optimizations, unless you count not repeating the ploidy getter.  @lbergelson am I missing something?

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #1684 and #1686.

### CODE REVIEW

**Comment by davidbenjamin** [comment]
@droazen do you mind reviewing this?

**Comment by akiezun** [comment]
@davidbenjamin can you make the title more descriptive? What tools are affected?

**Comment by davidbenjamin** [comment]
@akiezun There is no logic change, just things like using Java 8 to get one-liners and replacing argument-checking boilerplate with our existing utility methods.

**Comment by SHuang-Broad** [comment]
Hi @davidbenjamin and @droazen , I was actually looking into the errors mentioned in `InfiniteRandomMatingPopulationModel.java` in #1856 , which under this PR becomes `IndependentSampleGenotypesModel.java`.

I just want to bring this to attention, but do realize that @davidbenjamin 's PR doesn't change logic.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java [hunk 1]
```diff
@@ -114,13 +112,7 @@ private GenotypeAlleleCounts(final int ploidy, final int index, final int[] sort
...
     }
 
-    /**
-     * Returns the genotype's ploidy.
-     * @return 0 or greater.
-     */
-    public int ploidy() {
-        return ploidy;
-    }
+    public int ploidy() { return ploidy; }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: public methods should have javadoc, even if the javadoc is trivial

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java [hunk 2]
```diff
@@ -652,26 +628,17 @@ GenotypeAlleleCounts copy() {
...
-                return Collections.singletonList(allelesToUse.get(sortedAlleleCounts[0]));
-            } else {
-                return Collections.nCopies(ploidy, allelesToUse.get(sortedAlleleCounts[0]));
-            }
+            return Collections.nCopies(ploidy, allelesToUse.get(sortedAlleleCounts[0]));
         } else {
-            final Allele[] myAlleles = new Allele[ploidy];
-            int nextIndex = 0;
+            final List<T> result = new ArrayList<>(ploidy);
             for (int i = 0, ii = 0; i < distinctAlleleCount; i++) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: You have both `i` and `ii` in this loop -- recommend at least giving `ii` a more descriptive name to avoid confusion.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java [hunk 3]
```diff
@@ -652,26 +628,17 @@ GenotypeAlleleCounts copy() {
...
-                final Allele allele = allelesToUse.get(sortedAlleleCounts[ii++]);
+                final T allele = allelesToUse.get(sortedAlleleCounts[ii++]);
                 final int repeats = sortedAlleleCounts[ii++];
-                for (int j = 0; j < repeats; j++) {
-                    myAlleles[nextIndex++] = allele;
-                }
+                result.addAll(Collections.nCopies(repeats, allele));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: Using `Collections.nCopies()` is certainly cleaner, but adds an extra construction/garbage collection of a Collection object per iteration compared to the previous version. This is only worth caring about if this method is performance-sensitive, of course -- if it's not, then definitely prefer the cleaner version.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java [hunk 4]
```diff
@@ -249,6 +234,7 @@ protected VariantCallContext calculateGenotypes(final FeatureContext features,
 
         final OutputAlleleSubset outputAlternativeAlleles = calculateOutputAlleleSubset(AFresult);
 
+        // posterior probability that at least one alt allele exists in the samples
         final double PoFGT0 = Math.pow(10, AFresult.getLog10PosteriorOfAFGT0());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: Consider renaming this variable (`PoFGT0`) to something more human-readable/comprehensible like `postProbOfOneAltAllele`, assuming `PoFGT0` is not a universally-recognized standard abbreviation that I am just ignorant of :)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java [hunk 5]
```diff
@@ -339,27 +325,18 @@ private OutputAlleleSubset(final int count, final Allele[] alleles, final int[]
         }
 
         private List<Allele> outputAlleles(final Allele referenceAllele) {
-            final ArrayList<Allele> result = new ArrayList<>(count + 1);
-            result.add(referenceAllele);
-            for (int i = 0; i < count; i++) {
-                result.add(alleles[i]);
-            }
-            return result;
+            return Stream.concat(Stream.of(referenceAllele), Arrays.stream(alleles, 0, count)).collect(Collectors.toList());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: I'll just take this opportunity to say that we have no policy in GATK4 mandating that all loops be replaced with Java 8-style one-liners, and that, in my experience, sometimes the streaming version is clearer/more readable, and other times the old-school loop is clearer/more readable. There's no shame in a for loop -- we should always prefer the clearer/more straightforward version, whatever that may be.

Now, if you think the streaming version is more readable here, then I defer to your judgment despite personally thinking that the old-fashioned loop has the slight readability edge here :)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java [hunk 6]
```diff
@@ -339,27 +325,18 @@ private OutputAlleleSubset(final int count, final Allele[] alleles, final int[]
...
+            return Arrays.stream(mleCounts, 0, count).boxed().collect(Collectors.toList());
         }
     }
 
 
     /**
-     * Provided the exact mode computations it returns the appropiate subset of alleles that progress to genotyping.
-     * @param afcr the exact model calcualtion result.
+     * Provided the exact mode computations it returns the appropriate subset of alleles that progress to genotyping.
+     * @param afcr the exact model calculation result.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: Consider renaming `afcr` to something like `exactModelResult`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java [hunk 7]
```diff
@@ -603,34 +580,20 @@ protected final boolean passesCallThreshold(final double conf) {
...
-
-        final ArrayList<Double> MLEfrequencies = new ArrayList<>(alleleCountsofMLE.size());
-        // the MLEAC is allowed to be larger than the AN (e.g. in the case of all PLs being 0, the GT is ./. but the exact model may arbitrarily choose an AC>1)
-        for (final int AC : alleleCountsofMLE ) {
-            MLEfrequencies.add(Math.min(1.0, (double) AC / (double) AN));
-        }
-        return MLEfrequencies;
+    private List<Double> calculateMLEAlleleFrequencies(final List<Integer> alleleCountsofMLE, final GenotypesContext genotypes) {
+        final long AN = genotypes.stream().flatMap(g -> g.getAlleles().stream()).filter(Allele::isCalled).count();
+        return alleleCountsofMLE.stream().map(AC -> Math.min(1.0, (double) AC / AN)).collect(Collectors.toList());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: With dense streaming expressions like this I think it's good practice to include a one-line comment before each statement describing briefly what the intent is.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingLikelihoods.java [hunk 8]
```diff
@@ -35,17 +35,11 @@
...
         Utils.validateArg (ploidyModel.numberOfSamples() ==  likelihoods.size(), "there must be exactly one likelihood set for each sample");
-        if (likelihoods.stream().anyMatch(lik -> lik == null)){
-            throw new IllegalArgumentException("no genotype likelihood is allowed to be null");
-        }
-
-        this.likelihoods = likelihoods.toArray(new GenotypeLikelihoods[likelihoods.size()]);
-        this.alleles = alleles;
-        this.ploidyModel = ploidyModel;
+        this.likelihoods = Utils.nonNull(likelihoods, "the likelihood collection cannot be null").toArray(new GenotypeLikelihoods[likelihoods.size()]);
+        likelihoods.stream().forEach(lik -> Utils.nonNull(lik, "no genotype likelihood is allowed to be null"));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingLikelihoods.java]: Use the existing `Utils.containsNoNull()` method.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java [hunk 9]
```diff
@@ -17,25 +20,12 @@
...
-            }
-            ploidySum += p;
-        }
-        this.ploidySum = ploidySum;
-        isHomogeneous = ploidies.length == 0 || ploidies.length * this.ploidies[0] == ploidySum;
-        this.sampleList = sampleList;
+        this.sampleList = Utils.nonNull(sampleList, "the sample list cannot be null");
+        this.ploidies = Utils.nonNull(ploidies, "the ploidies cannot be null").clone();
+        Utils.validateArg(sampleList.numberOfSamples() == ploidies.length, "sample-list and ploidy array length must match");
+        Arrays.stream(ploidies).forEach(p -> {if (p < 0) throw new IllegalArgumentException("no ploidy can be less than 0");});
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java]: `p -> Utils.validateArg(p >= 0, "no ploidy can be less than 0")`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java [hunk 10]
```diff
@@ -17,25 +20,12 @@
...
+        ploidySum = (int) MathUtils.sum(ploidies);
+        isHomogeneous = ploidies.length == 0 || Arrays.stream(ploidies).allMatch(p -> p == ploidies[0]);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java]: Previous version was `ploidies.length == 0 || ploidies.length * this.ploidies[0] == ploidySum`, which would not necessarily have required all the ploidies to equal the first ploidy, as in the new version. Is this change intended?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import org.broadinstitute.hellbender.utils.genotyper.LikelihoodMatrix;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class delegates genotyping to allele count- and ploidy-dependent {@link GenotypeLikelihoodCalculator}s
+ * under the assumption that sample genotypes are independent conditional on their population frequencies.
+ */
+public final class IndependentSampleGenotypesModel {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: This class has had quite a makeover -- could any of the changes you made possibly affect `HaplotypeCaller` results? Have you run the `HaplotypeCallerIntegrationTest` in gatk-protected using a snapshot of this branch?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    private final int cacheAlleleCountCapacity;
+    private GenotypeLikelihoodCalculator[][] likelihoodCalculators;
+
+    public IndependentSampleGenotypesModel() { this(10,50); }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: Declare `10` and `50` as `static final` constants with descriptive names at the top of the class

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java (hunk 11 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: After actually looking at the new vs. old code I can see that things are most likely unchanged, but it still would be good to test.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        final AlleleListPermutation<A> permutation = data.permutation(genotypingAlleles);
+        final AlleleLikelihoodMatrixMapper<A> alleleLikelihoodMatrixMapper = AlleleLikelihoodMatrixMapper.newInstance(permutation);
+
+        final int sampleCount = data.numberOfSamples();
+        final PloidyModel ploidyModel = data.ploidyModel();
+        final List<GenotypeLikelihoods> genotypeLikelihoods = new ArrayList<>(sampleCount);
+        final int alleleCount = genotypingAlleles.numberOfAlleles();
+
+        GenotypeLikelihoodCalculator likelihoodsCalculator = sampleCount > 0 ? getLikelihoodsCalculator(ploidyModel.samplePloidy(0), alleleCount) : null;
+        for (int i = 0; i < sampleCount; i++) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: Do we have good, independent unit tests covering the four cases you collapsed into a single method (0 samples, 1 sample, multi-sample heterogeneous, and multi-sample homogeneous)?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    private GenotypeLikelihoodCalculator[][] likelihoodCalculators;
+
+    public IndependentSampleGenotypesModel() { this(10,50); }
+
+    public IndependentSampleGenotypesModel(final int calculatorCachePloidyCapacity, final int calculatorCacheAlleleCapacity) {
+        cacheAlleleCountCapacity = calculatorCacheAlleleCapacity;
+        likelihoodCalculators = new GenotypeLikelihoodCalculator[calculatorCachePloidyCapacity][calculatorCacheAlleleCapacity];
+    }
+
+    public <A extends Allele> GenotypingLikelihoods<A> calculateLikelihoods(final AlleleList<A> genotypingAlleles, final GenotypingData<A> data) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: Add javadoc for this public method

**Comment by droazen** [comment]
Review complete, back to @davidbenjamin, merge after addressing comments (no need for a second-pass review), but please do first run `HaplotypeCallerIntegrationTest` in protected with a snapshot of this branch to make sure there are no changes.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java (hunk 1 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java (hunk 2 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: Rewrote the loop to have only one index.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java (hunk 3 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeAlleleCounts.java]: Checked that this is not called in performance-sensitive code.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java (hunk 4 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java (hunk 5 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: I agree that one should be pragmatic.  For me personally the streaming one-liner is faster to parse in this case because there aren't extra lines for constructing and returning and because elements of the result are contiguous in the code.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java (hunk 6 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingLikelihoods.java (hunk 8 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingLikelihoods.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java (hunk 9 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java (hunk 10 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/HeterogeneousPloidyModel.java]: This is intentional.  Previous version was a bug.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java (hunk 11 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: will do.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java (hunk 12 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java (hunk 14 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: done

**Comment by davidbenjamin** [comment]
@SHuang-Broad I put your bug fix in the last commit.

**Comment by davidbenjamin** [comment]
@droazen This does not affect any of the integration tests or unit tests in gatk-protected, except insofar as `InfiniteRandomMatingPopulationModel` must be renamed.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java (hunk 11 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/IndependentSampleGenotypesModel.java]: done.
